pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java

```diff
@@ -53,102 +53,148 @@ public BytesOffHeapMutableDictionary(int estimatedCardinality, int maxOverflowHa
+
+    byte[] lowerValue;
+    if (lower.equals(RangePredicate.UNBOUNDED)) {
+      lowerValue = _min;
+    } else {
+      lowerValue = BytesUtils.toBytes(lower);
+    }
+    byte[] upperValue;
+    if (upper.equals(RangePredicate.UNBOUNDED)) {
+      upperValue = _max;
+    } else {
+      upperValue = BytesUtils.toBytes(upper);
+    }
+
+    if (includeLower && includeUpper) {
+      for (int dictId = 0; dictId < numValues; dictId++) {
+        byte[] value = getBytesValue(dictId);
+        if (ByteArray.compare(value, lowerValue) >= 0 && ByteArray.compare(value, upperValue) <= 0) {
+          dictIds.add(dictId);
+        }
+      }
+    } else if (includeLower) {
+      for (int dictId = 0; dictId < numValues; dictId++) {
+        byte[] value = getBytesValue(dictId);
+        if (ByteArray.compare(value, lowerValue) >= 0 && ByteArray.compare(value, upperValue) < 0) {
+          dictIds.add(dictId);
+        }
+      }
+    } else if (includeUpper) {
+      for (int dictId = 0; dictId < numValues; dictId++) {
+        byte[] value = getBytesValue(dictId);
+        if (ByteArray.compare(value, lowerValue) > 0 && ByteArray.compare(value, upperValue) <= 0) {
+          dictIds.add(dictId);
+        }
+      }
+    } else {
+      for (int dictId = 0; dictId < numValues; dictId++) {
+        byte[] value = getBytesValue(dictId);
+        if (ByteArray.compare(value, lowerValue) > 0 && ByteArray.compare(value, upperValue) < 0) {
+          dictIds.add(dictId);
+        }
+      }
+    }
+
+    return dictIds;
   }
 
   @Override
-  public void doClose()
-      throws IOException {
-    _byteStore.close();
+  public ByteArray getMinVal() {
+    return new ByteArray(_min);
   }
 
   @Override
-  protected void setRawValueAt(int dictId, Object rawValue, byte[] serializedValue) {
-    _byteStore.add(serializedValue);
+  public ByteArray getMaxVal() {
+    return new ByteArray(_max);
   }
 
   @Override
-  public void index(@Nonnull Object rawValue) {
-    byte[] bytes = BytesUtils.toBytes(rawValue);
-    ByteArray byteArray = new ByteArray(bytes);
-    indexValue(byteArray, bytes);
-    updateMinMax(byteArray);
+  public ByteArray[] getSortedValues() {
+    int numValues = length();
+    ByteArray[] sortedValues = new ByteArray[numValues];
+
+    for (int dictId = 0; dictId < numValues; dictId++) {
+      sortedValues[dictId] = new ByteArray(getBytesValue(dictId));
+    }
+
+    Arrays.sort(sortedValues);
+    return sortedValues;
   }
 
   @Override
-  public boolean inRange(@Nonnull String lower, @Nonnull String upper, int dictIdToCompare, boolean includeLower,
-      boolean includeUpper) {
-    return valueInRange(new ByteArray(BytesUtils.toBytes(lower)), new ByteArray(BytesUtils.toBytes(upper)),
-        includeLower, includeUpper, new ByteArray(getBytesValue(dictIdToCompare)));
+  public int indexOf(String stringValue) {
+    byte[] bytesValue = BytesUtils.toBytes(stringValue);
+    return getDictId(new ByteArray(bytesValue), bytesValue);
```

Comment:
Here you can just use `BytesUtils. toByteArray(stringValue)` which is already in this PR. 😉

Reply:
I need both byte[] and ByteArray, so I only fetched the byte[]

Reply:
Oh I see